src/procedures/proc_common_neighbors.c

```diff
@@ -0,0 +1,316 @@
+	GrB_Matrix At;              // Transposed relationship matrix.
+	GrB_Matrix SIM;             // Similarity matrix, SIM[i,j] = #routes between i and j.
+	GrB_Vector DEGREE;          // Out degree, DEGREE[i] = i's out degree.
+	GxB_Scalar thunk;           // Row offset.
+	GrB_Descriptor desc;        // Descriptor.
+	GxB_SelectOp select_op;     // Selects upper triangle.
+	GxB_MatrixTupleIter *iter;  // Matrix iterator.
+	int iteration;              // Iteration.
+	Node node1;                 // Node 1.
+	Node node2;                 // Node 2.
+	Graph *graph;               // Graph object.
+	SIValue *output;            // Array with 6 entries ["node1", node, "node2", node, "score", score].
+} SimilarityContext;
+
+// Create a new similarity context.
+static SimilarityContext *SimilarityContext_New() {
+	SimilarityContext *ctx = rm_malloc(sizeof(SimilarityContext));
+	ctx->nrows = 0;
+	ctx->batch_size = 0;
+	ctx->iter = NULL;
+	ctx->iteration = -1;
+	ctx->A = GrB_NULL;
+	ctx->At = GrB_NULL;
+	ctx->SIM = GrB_NULL;
+	ctx->desc = GrB_NULL;
+	ctx->thunk = GrB_NULL;
+	ctx->DEGREE = GrB_NULL;
+	ctx->select_op = GrB_NULL;
+	ctx->graph = QueryCtx_GetGraph();
+
+	// Setup reusable output.
+	ctx->output = array_new(SIValue, 6);
+	ctx->output = array_append(ctx->output, SI_ConstStringVal("node1"));
+	ctx->output = array_append(ctx->output, SI_Node(NULL)); // Place holder.
+	ctx->output = array_append(ctx->output, SI_ConstStringVal("node2"));
+	ctx->output = array_append(ctx->output, SI_Node(NULL)); // Place holder.
+	ctx->output = array_append(ctx->output, SI_ConstStringVal("score"));
+	ctx->output = array_append(ctx->output, SI_DoubleVal(0.0)); // Place holder.
+
+	return ctx;
+}
+
+// Free similarity context.
+static void SimilarityContext_Free(SimilarityContext *ctx) {
+	if(ctx) {
+		if(ctx->output) array_free(ctx->output);
+		if(ctx->At != GrB_NULL) GrB_free(&ctx->At);
+		if(ctx->SIM != GrB_NULL) GrB_free(&ctx->SIM);
+		if(ctx->desc != GrB_NULL) GrB_free(&ctx->desc);
+		if(ctx->thunk != GrB_NULL) GrB_free(&ctx->thunk);
+		if(ctx->DEGREE != GrB_NULL) GrB_free(&ctx->DEGREE);
+		if(ctx->select_op != GrB_NULL) GrB_free(&ctx->select_op);
+		if(ctx->iter) GxB_MatrixTupleIter_free(ctx->iter);
+		rm_free(ctx);
+	}
+}
+
+/* Selects the upper triangle portion of a matrix.
+ * taking into account row offset. */
+bool _select_triu(GrB_Index i, GrB_Index j, GrB_Index nrows, GrB_Index ncols, const void *x,
+				  const void *thunk) {
+
+	int64_t offset = *((int64_t *)thunk);
+	// Upper triangular.
+	return j > (offset + i);
+}
+
+// Extract row min through max from `A`, store result is `C`.
+static void _extract_rows(GrB_Matrix C, GrB_Matrix A, GrB_Index min, GrB_Index max) {
+	assert(C && A && (min <= max));
+
+	GrB_Info info;
+	GrB_Index nrows;
+	GrB_Index ncols;
+	GrB_Matrix_nrows(&nrows, A);
+	GrB_Matrix_ncols(&ncols, A);
+
+	max = MIN(max, nrows);
+
+	// Build array of row indices to extract.
+	GrB_Index count = max - min;
```

Comment:
I can't tell whether `nrows` is guaranteed to be >= `min`, as we might reassign `max` at 105. If `nrows < min`, this will overflow.